File: Python/306. Additive Number.py
class Solution:
    def find_rec(self, n1, n2, s, found):
            if s == "" and found:
                return True
            n3 = str(n1 + n2)
            idx = min(len(n3),len(s))
            if s[:idx] == n3:
                return self.find_rec(n2, int(n3), s[idx:], True)
            return False
        
    def isAdditiveNumber(self, num: str) -> bool:
        for i in range(1, len(num)-1):
            n1 = int(num[:i])
            if str(n1) != num[:i]:
                break
            for j in range(i+1, len(num)):
                n2 = int(num[i:j])
                if str(n2) != num[i:j]:
                    break
                found = self.find_rec(n1, n2, num[j:], False)
                if found:
                    return True
        return False

File: Java/306. Additive Number.java
import java.math.BigInteger;
class Solution {
    public boolean findrec(BigInteger n1,BigInteger n2,String s,Boolean found){
        if(found && s.length()==0){
            return true;
        }
        BigInteger sum=n1.add(n2);
        String n3=sum.toString();
        int idx=Math.min(s.length(),n3.length());
        String s2=s.substring(0,idx);
        if(s2.equals(n3)){
            return findrec(n2,sum,s.substring(idx),true);
        }

return false;
    }
    public boolean isAdditiveNumber(String num) {
        int n=num.length();
        for(int i=1;i<n-1;i++){
            BigInteger n1=new BigInteger(num.substring(0,i));
            if(!n1.toString().equals(num.substring(0,i))){
                break;
            }
            for(int j=i+1;j<n;j++){
                BigInteger n2=new BigInteger(num.substring(i,j));
                if(!n2.toString().equals(num.substring(i,j))){
                      break;
                  }
                if(findrec(n1,n2,num.substring(j),false)){
                    return true;
                }
            }
        }
 return false;
    }
}
